processing/src/main/java/org/apache/druid/segment/serde/ComplexMetricSerde.java

```diff
@@ -115,11 +115,13 @@ public Function<Object, Long> inputSizeFn()
    */
   public byte[] toBytes(@Nullable Object val)
   {
-    if (val != null) {
+    if (val == null) {
+      return ByteArrays.EMPTY_ARRAY;
+    } else if (val instanceof byte[]) {
+      return (byte[]) val;
+    } else {
```

Comment:
I think this is incorrect because there might exist a serde where the deserialized value is byte[], however, the serialized value can be something else. 

For example, different compression techniques can transform a deserialized byte[] array to a different serialized byte[] array. This would be unique to the complex type, so it shouldn't be a universal change. Instead, the following changes can be made:

1. If this only occurs in the HllSketchBuild serde, you can make this change specific to the HllSketchBuild serde. 
2. Else, if this occurs with all the complex types while running MSQ tasks on complex metrics, you can make the changes in MSQ-specific classes (after confirming where the byte[] array is coming from).

Reply:
Thanks for pointing that out! It looks like this issue might not be limited to just the HLLSketch. None of the ObjectStrategies seem to fit this scenario, as they all expect specific objects for the toBytes() method. Using it with a byte array directly would lead to the CCE. I'll look into why selector.getObject is returning a byte array for a real-time segment maybe it could be fixed there as well. It is not appears in MSQ classes directly tho, it shows up at ComplexFieldWriter.writeTo()

Reply:
np! Since the issue is bubbling up from the column value selectors, you can check who creates those, and why they are returning a byte array instead of the desired object type.

Reply:
I made some research and it looks like it's normal for selector.getObject to return a byte[] for real-time segments. This happens because it uses RowBasedColumnValueSelector instead of ObjectColumnSelector. It doesn't go through [GenericIndexed.copyBufferAndGet()](https://github.com/apache/druid/blob/169a8dbd1aa7bce86cb1508d18f1f342887f5020/processing/src/main/java/org/apache/druid/segment/data/GenericIndexed.java#L553) and strategy.fromByteBuffer(), and just grabs the byte[] directly from the ByteBuffer using [rowSupplier.get()](https://github.com/apache/druid/blob/169a8dbd1aa7bce86cb1508d18f1f342887f5020/processing/src/main/java/org/apache/druid/segment/RowBasedColumnSelectorFactory.java#L513) (at least in case of ComplexType from RealTime segment MSQ).

Right now, we don't have an ObjectStrategy that needs a byte[] as input, so initial setup would likely cause CCE failures. But you're right, we might need to handle this in the future. So,  I've moved this check to the ObjectStrategy as a defaul void, making it easy to override if needed.